flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierHandlerAction.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io.checkpointing;
+
+import org.apache.flink.runtime.checkpoint.CheckpointException;
+import org.apache.flink.runtime.checkpoint.channel.InputChannelInfo;
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+
+import java.io.IOException;
+
+/**
+ * Represents a state in a state machine of processing a checkpoint. There are 4 base states:
+ *
+ * <ul>
+ *   <li>Waiting for an aligned barrier
+ *   <li>Collecting aligned barriers
+ *   <li>Waiting for an unaligned barrier
+ *   <li>Collecting unaligned barriers
+ * </ul>
+ *
+ * <p>Additionally depending on the configuration we can switch between aligned and unaligned
+ * actions.
+ */
+interface BarrierHandlerAction {
+    BarrierHandlerAction alignmentTimeout(Context context, CheckpointBarrier checkpointBarrier)
+            throws IOException, CheckpointException;
+
+    BarrierHandlerAction announcementReceived(
+            Context context, InputChannelInfo channelInfo, int sequenceNumber) throws IOException;
+
+    BarrierHandlerAction barrierReceived(
+            Context context, InputChannelInfo channelInfo, CheckpointBarrier checkpointBarrier)
+            throws IOException, CheckpointException;
+
+    BarrierHandlerAction abort(long cancelledId) throws IOException;
+
+    /**
+     * An entry point for communication between {@link BarrierHandlerAction} and {@link
+     * SingleCheckpointBarrierHandler}.
+     */
+    interface Context {
+        boolean allBarriersReceived();
+
+        void triggerGlobalCheckpoint(CheckpointBarrier checkpointBarrier) throws IOException;
+
+        void triggerTaskCheckpoint(CheckpointBarrier checkpointBarrier) throws CheckpointException;
```

Comment:
sure, can rename them. However I don't like the `initCheckpoint` name. We should call it only for `unaligned` checkpoints to initialise checkpoints in input channels. `initCheckpoint` could indicate it should be called in case of aligned checkpoints as well.

How about `initInputsCheckpoint(ing)`?

Reply:
Sounds good. But in that case maybe we should also rename `SubtaskCheckpointCoordinatorImpl#initCheckpoint` to keep it consistent?

Reply:
Sure can do that.